kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/compose/compose.kt

```diff
@@ -0,0 +1,50 @@
+package io.kotest.matchers.compose
+
+import io.kotest.assertions.fail
+import io.kotest.matchers.Matcher
+import io.kotest.matchers.MatcherResult
+import kotlin.reflect.KProperty1
+
+/**
+ * Compose matchers along with the property to extract to test against.
+ * All matchers must be the same type `T`. If any matcher fails, composed matcher fails, too.
+ */
+@Suppress("UNCHECKED_CAST")
+fun <T : Any?> Matcher.Companion.compose(
+   vararg pairs: Pair<Matcher<*>, KProperty1<T, *>>
```

Comment:
There's nothing enforcing that the property extracts a type compatible with the matcher.

Reply:
actually we do 
```kotlin
         runCatching {
            (matcher as Matcher<Any?>).test(prop(value))
         }.onFailure {
            fail("Mismatching types in argument ${i + 1} for composed matcher:\n\n${it.message}")
         }.getOrThrow()
```
below

Reply:
I'll add additional test to prove it's working correctly